crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -36,68 +36,137 @@ impl FormatNodeRule<ExprBinOp> for FormatExprBinOp {
             range: _,
         } = item;
 
-        let should_break_right = self.parentheses == Some(Parentheses::Custom);
-
-        if should_break_right {
-            let left_group = f.group_id("BinaryLeft");
-
-            write!(
-                f,
-                [
-                    // Wrap the left in a group and gives it an id. The printer first breaks the
-                    // right side if `right` contains any line break because the printer breaks
-                    // sequences of groups from right to left.
-                    // Indents the left side if the group breaks.
-                    group(&format_args![
+        let layout = if self.parentheses == Some(Parentheses::Custom) {
+            BinaryLayout::from(item)
+        } else {
+            BinaryLayout::Default
+        };
+
+        match layout {
+            BinaryLayout::Default => {
+                let comments = f.context().comments().clone();
+                let operator_comments = comments.dangling_comments(item.as_any_node_ref());
+                let needs_space = !is_simple_power_expression(item);
+
+                let before_operator_space = if needs_space {
+                    soft_line_break_or_space()
+                } else {
+                    soft_line_break()
+                };
+
+                write!(
+                    f,
+                    [
+                        left.format(),
+                        before_operator_space,
+                        op.format(),
+                        trailing_comments(operator_comments),
+                    ]
+                )?;
+
+                // Format the operator on its own line if the right side has any leading comments.
+                if comments.has_leading_comments(right.as_ref()) {
+                    write!(f, [hard_line_break()])?;
+                } else if needs_space {
+                    write!(f, [space()])?;
+                }
+
+                write!(f, [group(&right.format())])
+            }
+
+            BinaryLayout::ExpandLeft => {
+                let parentheses_group = f.group_id("BinaryParentheses");
+                write!(
+                    f,
+                    [group(&format_args![
                         if_group_breaks(&text("(")),
-                        indent_if_group_breaks(
+                        // Only works thanks to the line break of `soft_block_indent`
+                        fits_expanded(&indent_if_group_breaks(
                             &format_args![
-                                soft_line_break(),
-                                left.format(),
-                                soft_line_break_or_space(),
-                                op.format(),
-                                space()
+                                if_group_breaks(&soft_line_break())
+                                    .with_group_id(Some(parentheses_group)),
+                                &group(&left.format())
```

Comment:
The problem is that this group may now be, incorrectly, printed in flat mode if it happens that the printer measures the most outer group and that happens to fit. This is because the printer only computes *fits* ones per line and, when true, assumes that all following groups up to the next line break fit.